File: backend-v2/app/schemas/pricing.py
from pydantic import BaseModel, ConfigDict, Field
from typing import Optional


class CompetitorInput(BaseModel):
    name: str
    discount_rate: float = Field(ge=0, le=100)


class PricingCalculateRequest(BaseModel):
    project_id: Optional[str] = None
    budget: float = Field(ge=0, default=0)
    ex_tax_price: Optional[float] = None
    inc_tax_price: Optional[float] = None
    tax_rate: float = Field(ge=0, le=1, default=0.06)
    profit_margin: float = Field(ge=0, default=15)
    risk_factor: int = Field(ge=1, le=5, default=3)
    pricing_method: str = Field(default="linear")
    k_value: float = Field(ge=0, le=100, default=95)
    sensitivity: float = Field(ge=0, default=2)
    effective_bidder_count: int = Field(ge=1, default=5)
    tech_score: float = Field(ge=0, le=100, default=75)
    competitors: list[CompetitorInput] = Field(default_factory=list)


class PricingBreakdown(BaseModel):
    ex_tax_price: float
    inc_tax_price: float
    ex_tax_cost: float
    inc_tax_cost: float
    discount_rate: float
    profit_margin: float
    tax_rate: float


class CompetitorResult(BaseModel):
    id: str
    name: str
    is_our: bool
    quote_price: float
    discount_rate: float
    price_score: float
    rank: int


class PricingScores(BaseModel):
    price_competitiveness: float
    profit_reasonability: float
    risk_controllability: float


class PricingCalculateResponse(BaseModel):
    success: bool = True
    message: str = ""
    breakdown: PricingBreakdown
    competitors: list[CompetitorResult]
    our_rank: int
    our_price_score: float
    benchmark_price: float
    min_review_price: float
    total_score: float
    scores: PricingScores
    ai_advice: str

File: frontend-v2/src/services/pricing.ts
import api from './api'

// ============ 请求类型 ============
export interface PricingPayload {
  project_id?: string
  budget?: number
  ex_tax_price?: number
  inc_tax_price?: number
  tax_rate?: number
  profit_margin?: number
  risk_factor?: number
  pricing_method?: string
  k_value?: number
  sensitivity?: number
  tech_score?: number
  competitors?: Array<{
    name: string
    discount_rate: number
  }>
}

// ============ 响应类型 ============
export interface PricingBreakdown {
  ex_tax_price: number
  inc_tax_price: number
  ex_tax_cost: number
  inc_tax_cost: number
  discount_rate: number
  profit_margin: number
  tax_rate: number
}

export interface CompetitorResult {
  id: string
  name: string
  is_our: boolean
  quote_price: number
  discount_rate: number
  price_score: number
  rank: number
}

export interface PricingScores {
  price_competitiveness: number
  profit_reasonability: number
  risk_controllability: number
}

export interface PricingResponse {
  success: boolean
  message: string
  breakdown: PricingBreakdown | null
  competitors: CompetitorResult[]
  our_rank: number
  our_price_score: number
  benchmark_price: number
  min_review_price: number
  total_score: number
  scores: PricingScores
  ai_advice: string
}

// ============ API 函数 ============

/**
 * 计算报价策略
 * @param payload 报价计算参数
 * @returns 计算结果包含报价明细、竞品模拟得分和总评分
 */
export async function calculatePricing(payload: PricingPayload): Promise<PricingResponse> {
  const res = await api.post<PricingResponse>('/pricing/calculate', payload)
  return res.data
}
